### ISSUE TITLE
Should HBox / VBox be classes instead of functions?

### ISSUE THREAD

**Issue by izaid** [ipython/ipywidgets#576]
I've been writing some custom widgets, which are largely simple classes that subclass from some other existing IPython widget. These mostly reuse the existing widgets, and the classes are for communication and saving state.

What confused me is that the HBox and VBox are not classes themselves, but functions. This meant when I tried to inherit from them, I was getting weird errors until I realized it was Box that I had to inherit from. Why don't we make HBox and VBox simple classes that inherit from Box? Or is there a reason not to do this?

**Comment by SylvainCorlay** [comment]
@izaid you probably want to take a look at the https://github.com/ipython/ipywidgets/blob/master/docs/source/examples/Flexbox%20Layout.ipynb example, which shows how to use the `layout` attribute of the `Box` widget to have richer layouts.

We expose the entire css flexbox spec described here:
https://css-tricks.com/snippets/css/a-guide-to-flexbox/

The only gotcha is that  when a property is also accessible via a shorthand property, we only expose the shorthand. For example:
    - ``flex-grow``, ``flex-shrink`` and ``flex-basis`` are bound to ``flex``.
    - ``flex-wrap`` and ``flex-direction`` are bound to ``flex-flow``.
    - ``margin-[top/bottom/left/right]`` values are bound to ``margin``, etc.

**Comment by SylvainCorlay** [comment]
The only reason we show HBox and VBox is to avoid people the pain of accessing the layout attribute just for the flex direction.

**Comment by izaid** [comment]
Thanks, @SylvainCorlay. That was a lot of useful information (actually), but my question was really much simpler.

Instead of the function HBox, Why not simply define the following?
```
class HBox(Box):
    def __init__(self, *pargs, **kwargs):
        Box.__init__(self, *pargs, **kwargs)
        self.layout.display = 'flex'
```

That way, it's consistent with the other Widget definitions and users can inherit from it. And do something analogous for VBox.

**Comment by SylvainCorlay** [comment]
That makes sense. What you could do instead of setting layout attributes in __init__ is overload the `Instance(Layout)` model attribute of `Box` to have different default constructor attributes.

**Comment by SylvainCorlay** [comment]
If you are making custom js widgets, you can check out the new cookiecutter project that helps you with the boilerplate. https://github.com/jupyter/widget-cookiecutter

**Comment by hainm** [comment]
I am myself  surprised too when trying to inherit from HBox.

**Comment by hainm** [comment]
but the layout is very helpful.

**Comment by dmadeka** [comment]
+1 on this as well, inheriting from HBox and VBox is useful

**Comment by SylvainCorlay** [comment]
Grumbles

**Comment by SylvainCorlay** [comment]
I would be ok with making HBox and VBox inherit from Box with overloading of defaults for `display` and `flex`... They are only instances of `Box` after all.

**Comment by SylvainCorlay** [comment]
Fixed in #765 and #767.
